File: api/BeSureApi/Models/PurchaseOrder.cs
namespace BeSureApi.Models
{
    public class PurchaseOrderCreate
    {
        public int DemandId { get; set; }
        public int TenantOfficeId { get; set; }
        public int PartId { get; set; }
        public int PartIndentRequestId { get; set; }
        public int VendorId { get; set; }
        public int? ShipToTenantOfficeInfoId { get; set; }
        public int? ShipToCustomerSiteId { get; set; }
        public int BillToTenantOfficeInfoId { get; set; }
        public string? Description { get; set; }
        public int? VendorBranchId { get; set; }
        public decimal CgstRate { get; set; }
        public decimal SgstRate { get; set; }
        public decimal IgstRate { get; set; }
        public int StockTypeId { get; set; }
        public decimal Price { get; set; }
    }

    public class ImprestPurchaseOrderCreate
    {
        public List<PartListData> PartList { get; set; }
        public int VendorId { get; set; }
        public int ShipToTenantOfficeInfoId { get; set; }
        public int BillToTenantOfficeInfoId { get; set; }
        public string? Description { get; set; }
        public int? VendorBranchId { get; set; }
    }

    public class BulkPurchaseOrderCreate
    {
        public List<DemandDetailList> PartList { get; set; }
        public int VendorId { get; set; }
        public int ShipToTenantOfficeInfoId { get; set; }
        public int BillToTenantOfficeInfoId { get; set; }
        public string? Description { get; set; }
        public int? VendorBranchId { get; set; }
    }

    public class PartListData
    {
        public int Id { get; set; }
        public string Description { get; set; }
        public int Quantity {  get; set; }
        public int StockTypeId  {  get; set; }
        public decimal Price {  get; set; }
        public decimal Sgst {  get; set; }
        public decimal Cgst {  get; set; }
        public decimal Igst {  get; set; }
    }
    public class DemandDetailList
    {
        public int PartId { get; set; }
        public int DemandId { get; set; }
        public string Description { get; set; }
        public int Quantity { get; set; }
        public int StockTypeId { get; set; }
        public decimal Price { get; set; }
        public decimal Sgst { get; set; }
        public decimal Cgst { get; set; }
        public decimal Igst { get; set; }
        public int PartIndentRequestId { get; set; }
    }

    public class PurchaseOrderList
    {
        public int Id { get; set; }
        public string PoNumber { get; set; }
        public DateTime PoDate { get; set; }
        public int VendorId { get; set; }
        public string Vendor {  get; set; }
        public string TenantOffice { get; set; }
    }
    public class PurchaseOrders
    {
        public int Id { get; set; }
        public string PoNumber { get; set; }
        public DateTime PoDate { get; set; }
        public int VendorId { get; set; }
        public string Vendor { get; set; }
        public string TenantOffice { get; set; }
        public string PoStatus { get; set; }
    }

    public class PurchaseOrderDetail
    {
        public int Id { get; set; }
        public string PoNumber { get; set; }
        public DateTime PoDate { get; set; }
        public string Vendor { get; set; }
        public string TenantOffice { get; set; }
        public string BillToTenantOffice { get; set; }
        public string BillToAddress { get; set; }
        public string ShipToTenantOffice { get; set; }
        public string ShipToAddress { get; set; }
        public string? IndentRequestNumber { get; set; }
        public string PoStatus { get; set; }
        public decimal Quantity { get; set; }
        public decimal GrnReceivedQuantity { get; set; }
        public decimal Price { get; set; }
        public string? PartName { get; set; }
        public decimal CgstRate { get; set; }
        public decimal IgstRate { get; set; }
        public decimal SgstRate { get; set; }
        public string PoPartType { get; set; }
    }
}

File: ui/src/types/purchaseorder.ts
import { Decoder, array, boolean, nullable, number, object, string } from "decoders";

export interface RequestPOResponse {
    IsUpdated: Boolean;
}

export const requestPoResponseDecoder: Decoder<RequestPOResponse> = object({
    IsUpdated: boolean,
});

export interface RequestForPurchaseOrder {
    Id: number;
    VendorId: number;
    VendorTypeId: number | null;
    Price: number;
    StockTypeId: number;
    WarrantyPeriod: number;
}

export interface CreatePurchaseOrder {
    PartId: number;
    DemandId: number;
    PartIndentRequestId: number;
    TenantOfficeId: number;
    VendorId: number;
    VendorTypeId: number|null;
    VendorBranchId: number | null;
    ShipToTenantOfficeInfoId: number | null;
    ShipToCustomerSiteId: number | null;
    BillToTenantOfficeInfoId: number;
    Description: string | null;
    CgstRate: number;
    SgstRate: number;
    IgstRate: number;
    StockTypeId: number,
    Price: number;
}

export interface CreatePOResponse {
    IsPOCreated: Boolean;
}

export const createPoResponseDecoder: Decoder<CreatePOResponse> = object({
    IsPOCreated: boolean,
});

export interface PurchaseOrdersList {
    Id: number;
    PoNumber: string;
    PoDate: string;
    Vendor: string;
    VendorId: number;
    TenantOffice: string;
}

export const purchaseOrdersListDecoder: Decoder<PurchaseOrdersList> = object({
    Id: number,
    PoNumber: string,
    PoDate: string,
    Vendor: string,
    VendorId: number,
    TenantOffice: string
});

export interface LocationWisePurchaseOrdersList {
    PurchaseOrders: PurchaseOrdersList[],
    TotalRows: number;
    PerPage: number;
}

export const locationWisePurchaseOrdersListDecoder: Decoder<LocationWisePurchaseOrdersList> = object({
    PurchaseOrders: array(purchaseOrdersListDecoder),
    TotalRows: number,
    PerPage: number
});

export interface PurchaseOrderDetail {
    Id: number;
    PoNumber: string;
    PoDate: string;
    Vendor: string;
    TenantOffice: string;
    BillToTenantOffice: string;
    BillToAddress: string;
    ShipToTenantOffice: string;
    ShipToAddress: string;
    IndentRequestNumber: string | null;
    PoStatus: string;
    Quantity: number;
    GrnReceivedQuantity: number;
    Price: number;
    PartName: string | null;
    CgstRate: number;
    IgstRate: number;
    SgstRate: number;
    PoPartType: string;
}

export const purchaseOrderDecoder: Decoder<PurchaseOrderDetail> = object({
    Id: number,
    PoNumber: string,
    PoDate: string,
    Vendor: string,
    TenantOffice: string,
    BillToTenantOffice: string,
    BillToAddress: string,
    ShipToTenantOffice: string,
    ShipToAddress: string,
    IndentRequestNumber: nullable(string),
    PoStatus: string,
    Quantity: number,
    GrnReceivedQuantity: number,
    Price: number,
    PartName: nullable(string),
    CgstRate: number,
    IgstRate: number,
    SgstRate: number,
    PoPartType: string
})

export interface purchaseOrderDetails {
    PurchaseOrderDetails: PurchaseOrderDetail[],
}

export const purchaseOrderDetailsDecoder: Decoder<purchaseOrderDetails> = object({
    PurchaseOrderDetails: array(purchaseOrderDecoder),
});

export interface PurchaseOrders {
    Id: number;
    PoNumber: string;
    PoDate: string;
    Vendor: string;
    VendorId: number;
    TenantOffice: string;
    PoStatus: string;
}

export const purchaseordersDecoder: Decoder<PurchaseOrders> = object({
    Id: number,
    PoNumber: string,
    PoDate: string,
    Vendor: string,
    VendorId: number,
    TenantOffice: string,
    PoStatus: string
});

export interface ListPurchaseOrders {
    PurchaseOrders: PurchaseOrders[],
    TotalRows: number;
    PerPage: number;
}

export const ListPurchaseOrdersDecoder: Decoder<ListPurchaseOrders> = object({
    PurchaseOrders: array(purchaseordersDecoder),
    TotalRows: number,
    PerPage: number
});

export interface Select {
    value: any,
    label: any,
    code?: any
}

export interface SelectDetails {
    Select: Select[]
}